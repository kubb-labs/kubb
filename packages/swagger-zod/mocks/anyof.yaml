info:
  title: anyof test cases
  version: 1.0.0
openapi: 3.1.0
paths: {}
components:
  schemas:
    test:
      anyOf:
        - type: object
          properties:
            propertyA:
              type: string
          required:
            - propertyA
        - type: object
          properties:
            propertyA:
              type: string
            propertyB:
              type: string
          required:
            - propertyA
            - propertyB
