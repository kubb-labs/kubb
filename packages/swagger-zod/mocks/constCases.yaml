info:
  title: const test cases
  version: 1.0.0
openapi: 3.1.0
paths: {}
components:
  schemas:
    UuidSchema:
      type: string
      format: uuid
    NullableString:
      type:
        - string
        - "null"
    NullableStringWithAnyOf:
      anyOf:
        - type: string
        - type: "null"
    NullableStringUuid:
      type:
        - string
        - "null"
      format: uuid
    NullConst:
      required: true
      const: null
      description: |
        The value of this keyword MAY be null.
        Use of this keyword is functionally equivalent to an "enum" (Section 6.1.2) with a single value.
        An instance validates successfully against this keyword if its value is equal to the value of the keyword.
    StringValueConst:
      description: its value is equal to the value of the keyword
      required:
        - foobar
      properties:
        foobar:
          const: "foobar"
    NumberValueConst:
      description: its value is equal to the value of the keyword
      required:
        - foobar
      properties:
        foobar:
          const: 42
    MixedValueTypeConst:
      description: This probably should fail miserably
      required:
      - foobar
      properties:
        foobar:
          const: "foobar"
          type: number
