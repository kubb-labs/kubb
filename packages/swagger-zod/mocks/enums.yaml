definitions:
  enumVarNames.Type:
    enum:
    - 0
    - 1
    type: integer
    x-enum-varnames:
    - Pending
    - Received
  enumNames.Type:
    enum:
    - 0
    - 1
    x-enumNames:
      - Pending
      - Received
swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumVarNames.Type'
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNames.Type'
