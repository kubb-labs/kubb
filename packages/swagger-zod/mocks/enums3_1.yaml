openapi: 3.1.0
info:
  title: test
  version: "1.0.0"

components:
  schemas:
    enumNullable:
      type:
       - "string"
       - "null"
      enum:
      - Pending
      - Received

paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNullable'
