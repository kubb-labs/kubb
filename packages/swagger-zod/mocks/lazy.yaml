---
openapi: 3.0.1
info:
  title: Example Schema
servers:
- url: http://localhost/
paths:
  "/example":
    get:
      responses:
        '200':
          description: Ok
          content:
            application/json:
              schema:
                "$ref": "#/components/schemas/Example"
components:
  schemas:
    Example:
      type: object
      properties:
        nestedExamples:
          "$ref": "#/components/schemas/Example"
