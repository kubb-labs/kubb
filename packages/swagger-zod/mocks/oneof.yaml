info:
  title: oneof test cases
  version: 1.0.0
openapi: 3.1.0
paths: {}
components:
  schemas:
    test:
      oneOf:
        - type: object
          properties:
            propertyA:
              type: string
        - type: object
          properties:
            propertyA:
              type: string
