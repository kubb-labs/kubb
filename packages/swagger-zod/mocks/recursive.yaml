---
openapi: 3.0.1
info:
  title: Example Schema
servers:
- url: http://localhost/
paths:
  "/example":
    get:
      responses:
        '200':
          description: Ok
          content:
            application/json:
              schema:
                "$ref": "#/components/schemas/Example"
components:
  schemas:
    Example:
      type: object
      required: ['name', 'children']
      properties:
        name:
          type: string
        children:
          type: array
          items:
            $ref: "#/components/schemas/Example"
