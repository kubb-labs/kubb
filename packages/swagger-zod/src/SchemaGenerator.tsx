import { createRoot, File, useFile, usePlugin, usePluginManager } from '@kubb/react'
import { SchemaGenerator as Generator } from '@kubb/swagger'
import { Oas, Schema } from '@kubb/swagger/components'
import { useSchema } from '@kubb/swagger/hooks'
import { pluginKey as swaggerTypeScriptPluginKey } from '@kubb/swagger-ts'

import { pluginKey } from './plugin.ts'
import { zodParser } from './zodParser.tsx'

import type { AppContextProps } from '@kubb/react'
import type { Schema as SchemaType, SchemaGeneratorBuildOptions, SchemaMethodResult } from '@kubb/swagger'
import type { SchemaObject } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

function SchemaImports() {
  const plugin = usePlugin<PluginOptions>()
  const pluginManager = usePluginManager()
  const { path: root } = useFile()
  const { name } = useSchema()

  const withTypeAnnotation = plugin.options.typed

  // used for this.options.typed
  const typeName = pluginManager.resolveName({
    name,
    pluginKey: swaggerTypeScriptPluginKey,
    type: 'type',
  })
  const typeFileName = pluginManager.resolveName({
    name: name,
    pluginKey: swaggerTypeScriptPluginKey,
    type: 'file',
  })
  const typePath = pluginManager.resolvePath({
    baseName: typeFileName,
    pluginKey: swaggerTypeScriptPluginKey,
  })

  return (
    <>
      <File.Import name={['z']} path="zod" />
      {withTypeAnnotation && typeName && typePath && <File.Import isTypeOnly root={root} path={typePath} name={[typeName]} />}
    </>
  )
}

export class SchemaGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async schema(name: string, object: SchemaObject): SchemaMethodResult<FileMeta> {
    const { oas, pluginManager, mode, plugin, output } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    root.render(
      <Oas oas={oas}>
        <Oas.Schema generator={this} name={name} object={object}>
          <Schema.File output={output} mode={mode}>
            <SchemaImports />
          </Schema.File>
        </Oas.Schema>
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  getSource(name: string, schemas: SchemaType[], { keysToOmit, operation, description }: SchemaGeneratorBuildOptions & { required?: boolean } = {}): string[] {
    const texts: string[] = []

    // all checks are also inside this.schema(React)
    const withTypeAnnotation = this.options.typed && !operation

    // used for this.options.typed
    const typeName = this.context.pluginManager.resolveName({
      name,
      pluginKey: swaggerTypeScriptPluginKey,
      type: 'type',
    })

    const output = zodParser(schemas, {
      keysToOmit,
      name: this.context.pluginManager.resolveName({
        name: name,
        pluginKey,
        type: 'function',
      }),
      description,
      typeName: withTypeAnnotation ? typeName : undefined,
    })

    texts.push(output)

    return texts
  }
  /**
   * @deprecated only used for testing
   */
  buildSource(name: string, schema: SchemaObject, options: SchemaGeneratorBuildOptions = {}): string[] {
    const schemas = this.buildSchemas(schema, name)

    // all checks are also inside this.schema(React)
    // hack so Params will be optional when needed
    const required = Array.isArray(schema?.required) ? !!schema.required.length : !!schema?.required

    return this.getSource(name, schemas, { ...options, required })
  }
}
