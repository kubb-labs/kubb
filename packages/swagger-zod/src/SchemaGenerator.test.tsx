import path from 'node:path'

import { mockedPluginManager } from '@kubb/core/mocks'
import { OasManager } from '@kubb/swagger'

import { SchemaGenerator } from './SchemaGenerator.tsx'

import type { Plugin } from '@kubb/core'
import type { SchemaObject } from '@kubb/swagger/oas'
import type { PluginOptions } from './types.ts'

describe('Zod SchemaGenerator PetStore', async () => {
  const petStorePath = path.resolve(__dirname, '../mocks/petStore.yaml')
  const oas = await new OasManager().parse(petStorePath)

  test('generate schema for Pet', async () => {
    const generator = new SchemaGenerator(
      {
        dateType: 'string',
        include: undefined,
        transformers: {},
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pet', schemas?.Pet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate schema for Pets', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pets', schemas?.Pets as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate schema for OptionalPet', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('OptionalPet', schemas?.OptionalPet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate schema for OptionalPet typed', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        typed: true,
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('OptionalPet', schemas?.OptionalPet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate schema for PetArray', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('PetArray', schemas?.PetArray as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('ZodGenerator constCases', async () => {
  const discriminatorPath = path.resolve(__dirname, '../mocks/constCases.yaml')
  const oas = await new OasManager().parse(discriminatorPath)
  const generator = new SchemaGenerator(
    {
      transformers: {},
      dateType: 'string',
      unknownType: 'any',
      exclude: undefined,
      override: undefined,
      typed: false,
      include: undefined,
    },
    {
      oas,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      include: undefined,
    },
  )
  // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
  const schemas = oas.getDefinition().components?.schemas!

  test('UuidSchema generates a string with uuid format constraint', async () => {
    const schema = schemas['UuidSchema'] as SchemaObject
    const node = generator.buildSource('UuidSchema', schema)

    expect(node).toMatchSnapshot()
  })

  test('NullableString zodifies correctly', async () => {
    const schema = schemas?.['NullableString'] as SchemaObject
    const node = generator.buildSource('NullableString', schema)

    expect(node).toMatchSnapshot()
  })

  test('NullableStringWithAnyOf results in union of string and null', async () => {
    const schema = schemas['NullableStringWithAnyOf'] as SchemaObject
    const node = generator.buildSource('NullableStringWithAnyOf', schema)

    expect(node).toMatchSnapshot()
  })

  test('NullableStringUuid zodifies correctly to a uuid or null', async () => {
    const schema = schemas['NullableStringUuid'] as SchemaObject
    const node = generator.buildSource('NullableStringUuid', schema)

    expect(node).toMatchSnapshot()
  })

  test('NullConst zodifies correctly', async () => {
    const schema = schemas['NullConst'] as SchemaObject
    const node = generator.buildSource('NullConst', schema)

    expect(node).toMatchSnapshot()
  })

  test('StringValueConst correctly generates zod literal', async () => {
    const schema = schemas['StringValueConst'] as SchemaObject
    const node = generator.buildSource('StringValueConst', schema)

    expect(node).toMatchSnapshot()
  })

  test('NumberValueConst correctly generates zod literal', async () => {
    const schema = schemas['NumberValueConst'] as SchemaObject
    const node = generator.buildSource('NumberValueConst', schema)

    expect(node).toMatchSnapshot()
  })

  test('MixedValueTypeConst generates zod literal value correctly, overriding the type constraint', async () => {
    const schema = schemas['MixedValueTypeConst'] as SchemaObject
    const node = generator.buildSource('MixedValueTypeConst', schema)

    expect(node).toMatchSnapshot()
  })
})

describe('Zod SchemaGenerator lazy', async () => {
  const petStorePath = path.resolve(__dirname, '../mocks/lazy.yaml')
  const oas = await new OasManager().parse(petStorePath)

  test('generate schema for Example', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Example', schemas?.Example as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('Zod SchemaGenerator enums', async () => {
  const schemaPath = path.resolve(__dirname, '../mocks/enums.yaml')
  const oas = await new OasManager().parse(schemaPath)
  const generator = new SchemaGenerator(
    {
      transformers: {},
      dateType: 'string',
      unknownType: 'any',
      exclude: undefined,
      override: undefined,
      typed: false,
      include: undefined,
    },
    {
      oas,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      include: undefined,
    },
  )

  const schemas = oas.getDefinition().components?.schemas

  test('generate x-enum-varnames types', async () => {
    const node = generator.buildSource('enumVarNames', schemas?.['enumVarNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate x-enumNames types', async () => {
    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('Zod SchemaGenerator recursive', async () => {
  const petStorePath = path.resolve(__dirname, '../mocks/recursive.yaml')
  const oas = await new OasManager().parse(petStorePath)

  test('generate schema for Example', async () => {
    const generator = new SchemaGenerator(
      {
        transformers: {},
        dateType: 'string',
        unknownType: 'any',
        exclude: undefined,
        override: undefined,
        typed: false,
        include: undefined,
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Example', schemas?.Example as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('Zod SchemaGenerator anyof', async () => {
  const discriminatorPath = path.resolve(__dirname, '../mocks/anyof.yaml')
  const oas = await new OasManager().parse(discriminatorPath)
  const generator = new SchemaGenerator(
    {
      transformers: {},
      dateType: 'string',
      unknownType: 'any',
      exclude: undefined,
      override: undefined,
      typed: false,
      include: undefined,
    },
    {
      oas,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      include: undefined,
    },
  )
  // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
  const schemas = oas.getDefinition().components?.schemas!

  test('anyof with 2 objects', async () => {
    const schema = schemas['test'] as SchemaObject
    const node = generator.buildSource('test', schema)

    expect(node).toMatchSnapshot()
  })
})
