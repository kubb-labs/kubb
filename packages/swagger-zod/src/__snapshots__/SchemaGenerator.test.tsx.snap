// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Zod SchemaGenerator > 'MixedValueTypeConst' 1`] = `
"export const mixedValueTypeConst = z.object({ foobar: z.literal('foobar') }).describe('This probably should fail miserably')
"
`;

exports[`Zod SchemaGenerator > 'NullConst' 1`] = `
"export const nullConst = z.string().uuid().nullable()
"
`;

exports[`Zod SchemaGenerator > 'NullableString' 1`] = `
"export const nullableString = z.string().nullable()
"
`;

exports[`Zod SchemaGenerator > 'NullableStringUuid' 1`] = `
"export const nullableStringUuid = z.string().uuid().nullable()
"
`;

exports[`Zod SchemaGenerator > 'NullableStringWithAnyOf' 1`] = `
"export const nullableStringWithAnyOf = z.union([z.string(), z.null()])
"
`;

exports[`Zod SchemaGenerator > 'NumberValueConst' 1`] = `
"export const numberValueConst = z.object({ foobar: z.literal(42) }).describe('its value is equal to the value of the keyword')
"
`;

exports[`Zod SchemaGenerator > 'OptionalPet typed' 1`] = `
"export const optionalPetTyped = z.object({ id: z.number().optional(), name: z.string().optional(), tag: z.string().optional() }) as z.ZodType<OptionalPetTyped>
"
`;

exports[`Zod SchemaGenerator > 'OptionalPet' 1`] = `
"export const optionalPet = z.object({ id: z.number().optional(), name: z.string().optional(), tag: z.string().optional() }) as z.ZodType<OptionalPet>
"
`;

exports[`Zod SchemaGenerator > 'Order dateType date' 1`] = `
"export const orderDateTypeDate = z.object({
  id: z.number().optional(),
  petId: z.number().optional(),
  quantity: z.number().optional(),
  shipDate: z.date().optional(),
  status: z.enum(['placed', 'approved', 'delivered']).describe('Order Status').optional(),
  complete: z.boolean().optional(),
})
"
`;

exports[`Zod SchemaGenerator > 'Order dateType false' 1`] = `
"export const orderDateTypeFalse = z.object({
  id: z.number().optional(),
  petId: z.number().optional(),
  quantity: z.number().optional(),
  shipDate: z.string().optional(),
  status: z.enum(['placed', 'approved', 'delivered']).describe('Order Status').optional(),
  complete: z.boolean().optional(),
})
"
`;

exports[`Zod SchemaGenerator > 'Order dateType string' 1`] = `
"export const orderDateTypeString = z.object({
  id: z.number().optional(),
  petId: z.number().optional(),
  quantity: z.number().optional(),
  shipDate: z.string().datetime().optional(),
  status: z.enum(['placed', 'approved', 'delivered']).describe('Order Status').optional(),
  complete: z.boolean().optional(),
})
"
`;

exports[`Zod SchemaGenerator > 'Pet' 1`] = `
"export const pet = z.object({ id: z.number(), name: z.string(), date: z.date().optional(), tag: z.string().min(5).max(100).optional() })
"
`;

exports[`Zod SchemaGenerator > 'PetArray' 1`] = `
"export const petArray = z
  .array(z.lazy(() => Pet))
  .min(1)
  .max(3)
  .describe('List of Pet object')
"
`;

exports[`Zod SchemaGenerator > 'Pets' 1`] = `
"export const pets = z.array(z.object({ id: z.number(), name: z.string(), tag: z.string().optional() }))
"
`;

exports[`Zod SchemaGenerator > 'StringValueConst' 1`] = `
"export const stringValueConst = z.object({ foobar: z.literal('foobar') }).describe('its value is equal to the value of the keyword')
"
`;

exports[`Zod SchemaGenerator > 'UuidSchema' 1`] = `
"export const uuidSchema = z.string().uuid()
"
`;

exports[`Zod SchemaGenerator > 'anyof' 1`] = `
"export const anyof = z.union([z.object({ propertyA: z.string() }).strict(), z.object({ propertyA: z.string(), propertyB: z.string() }).strict()])
"
`;

exports[`Zod SchemaGenerator > 'enumNames' 1`] = `
"export const enumNames = z.enum(['0', '1'])
"
`;

exports[`Zod SchemaGenerator > 'enumNullable 3.1' 1`] = `
"export const enumNullable31 = z.enum(['Pending', 'Received']).nullable()
"
`;

exports[`Zod SchemaGenerator > 'enumVarNames' 1`] = `
"export const enumVarNames = z.union([z.literal(0), z.literal(1)])
"
`;

exports[`Zod SchemaGenerator > 'lazy' 1`] = `
"export const lazy = z.object({ nestedExamples: z.lazy(() => Example).optional() })
"
`;

exports[`Zod SchemaGenerator > 'recursive' 1`] = `
"export const recursive = z.object({ name: z.string(), children: z.array(z.lazy(() => Example)) })
"
`;
